### PR TITLE
NOISSUE - Fix domains bugs and feature to disable user self register

### LINKED ISSUE #55: Duplicate check for entities
**BUG REPORT**

1. What were you trying to achieve?
To create a group with the same name in different domains
3. What are the expected results?
I should be able to create groups with the same name in different domains
4. What are the received results?
Conflict error. Entity already exists
5. What are the steps to reproduce the issue?

- Create a group in a domain  e.g group1
- Log in to another domain and try create a group with the same name.

6. In what environment did you encounter the issue?
Magistrala Domains
7. Additional information you deem important:
It seems as though the conflict check is not done based on the domain but based on all the entities in the database. But this should be done on the domain level.

### LINKED ISSUE #60: List Domains not showing up member relations by default
**BUG REPORT**

1. What were you trying to achieve?

2. What are the expected results?

3. What are the received results?

4. What are the steps to reproduce the issue?

5. In what environment did you encounter the issue?

6. Additional information you deem important:

### LINKED ISSUE #72: Users Service: Error in Refresh & Issue Token logging Middleware functions
**BUG REPORT**

1. What were you trying to achieve?
Get new token from  Refresh Token endpoint

2. What are the expected results?
`failed to perform authentication over the entity : could not verify message using any of the signatures or keys`

3. What are the received results?
```html
<html>

<head>
	<title>502 Bad Gateway</title>
</head>

<body>
	<center>
		<h1>502 Bad Gateway</h1>
	</center>
	<hr>
	<center>nginx/1.23.3</center>
</body>

</html>
```
4. What are the steps to reproduce the issue?
Add random character at end of refresh token and sent request to refresh token endpoint

5. In what environment did you encounter the issue?
Docker, Linux

6. Additional information you deem important:
**Error Description:**
The issue arises as the token `t` is referenced within the defer block before being assigned its value from the service method calls. This cause error  `runtime error: invalid memory address or nil pointer dereference`

Existing Code :  https://github.com/absmach/magistrala/blob/main/users/api/logging.go#L59-L69
```go 
func (lm *loggingMiddleware) RefreshToken(ctx context.Context, refreshToken, domainID string) (t *magistrala.Token, err error) {
	defer func(begin time.Time) {
		message := fmt.Sprintf("Method refresh_token of type %s for refresh token %s took %s to complete", t.AccessType, refreshToken, time.Since(begin))
		if err != nil {
			lm.logger.Warn(fmt.Sprintf("%s with error: %s.", message, err))
			return
		}
		lm.logger.Info(fmt.Sprintf("%s without errors.", message))
	}(time.Now())
	return lm.svc.RefreshToken(ctx, refreshToken, domainID)
}
```

Can we have some thing like below?: 
```go
func (lm *loggingMiddleware) RefreshToken(ctx context.Context, refreshToken, domainID string) (t *magistrala.Token, err error) {
	startTime := time.Now()
	t, err = lm.svc.RefreshToken(ctx, refreshToken, domainID)
	message := fmt.Sprintf("Method refresh_token of type %s for refresh token %s took %s to complete", t.AccessType, refreshToken, time.Since(startTime))
	if err != nil {
		lm.logger.Warn(fmt.Sprintf("%s with error: %s.", message, err))
		return nil, err
	}
	lm.logger.Info(fmt.Sprintf("%s without errors.", message))
	return t, nil
}
```
Same we kind of error might occure also in Issue Token logging Middleware 
https://github.com/absmach/magistrala/blob/main/users/api/logging.go#L45-L55

### LINKED ISSUE #78: Duplicate items in list domains response
**BUG REPORT**

1. What were you trying to achieve?
Listing domains 

2. What are the expected results?
List of domains 

3. What are the received results?
List of domains with duplicate items 

4. What are the steps to reproduce the issue?
Logging with super admin credentials 
and send request to list domains endpoint

5. In what environment did you encounter the issue?
Docker, Linux

6. Additional information you deem important:

### PR DESCRIPTION
**Pull request by arvindh123**
Pull request title should be `MG-XXX - description` or `NOISSUE - description` where XXX is ID of issue that this PR relate to.
Please review the [CONTRIBUTING.md](https://github.com/absmach/magistrala/blob/master/CONTRIBUTING.md) file for detailed contributing guidelines.

### What does this do?
- Fix: Update Nginx entrypoint.sh file shebang
- Fix: Include membership users by default in listing domain users , added permission=membership as default query params for domain users listing (#60)
- Fix: Resolve conflict when creating identical thing names, thing keys, and group names in different domains (#55)
- Add: Implement a feature to disable user self-registration; users can only be added by a super admin
- Fix:  Duplicate items in list domains response #78
- Fix: #72 

### Which issue(s) does this PR fix/relate to?
Put here `Resolves #XXX` to auto-close the issue that your PR fixes (if such)
- Resolves #60
- Resolves #55 
- Resolves #78 
- Resolves #72

### CODE REVIEW

File: users/service.go [hunk 1]
```diff
@@ -503,7 +503,7 @@ func (svc *service) authorize(ctx context.Context, subjType, subjKind, subj, per
 func (svc service) Identify(ctx context.Context, token string) (string, error) {
 	user, err := svc.auth.Identify(ctx, &magistrala.IdentityReq{Token: token})
 	if err != nil {
-		return "", err
+		return "", errors.Wrap(errors.ErrAuthentication, err)
```
Comment (rodneyosodo) [inline users/service.go:506]: This creates a duplicate since Identify returns `errors.ErrAuthentication`

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: users/service.go (hunk 1 above)
Comment (arvindh123) [reply users/service.go:506]: `svc.auth.Identify` returns rpc errors , so `api.EncodeError` could not able to encode error and returns `500` error
Regrading rpc errors raised issue here #50

**Review by rodneyosodo — APPROVED** [review]
LGTM

**Review by ianmuchyri — APPROVED** [review]

**Comment by ianmuchyri** [comment]
@arvindh123 when I try listing entities who's list is empty it returns an error instead of an empty list

`2023-11-24 15:42:26 {"level":"warn","message":"Method list_things using token eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJkb21haW4iOiI1Y2Q3NDhkNy1iZjFmLTRjYTktYjEwYy00ZmZlM2Y0Mzc5YmEiLCJleHAiOjE3MDA4MzMxNTIsImlhdCI6MTcwMDgyOTU1MiwiaXNzIjoibWFnaXN0cmFsYS5hdXRoIiwic3ViIjoiYzIxN2NlYWYtOTJkYy00YmI1LWFlNDgtM2RjZjI3NTU5MzhkIiwidHlwZSI6MCwidXNlciI6ImMyMTdjZWFmLTkyZGMtNGJiNS1hZTQ4LTNkY2YyNzU1OTM4ZCJ9.M-swDOOUblSrryn3gAZmyYLW5p5EvXKgxWDYHCR4JQKKT7VBfA9nk7W8nxLw0SEHVTw90kIrq2AzuRz3gDqCdQ took 6.221668ms to complete with error: entity not found : entity not found : rpc error: code = Internal desc = entity not found.","ts":"2023-11-24T12:42:26.601743534Z"}`

File: auth/postgres/domains.go [hunk 2]
```diff
@@ -159,6 +159,11 @@ func (repo domainRepo) ListDomains(ctx context.Context, pm auth.Page) (auth.Doma
 	JOIN policies pc
 	ON pc.object_id = d.id`
 
+	if pm.SubjectID == "" {
```
Comment (dborovcanin) [inline auth/postgres/domains.go:162]: Please add a comment explaining this approach.

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**